yarn-project/noir-protocol-circuits/src/crates/private-kernel-lib/src/private_kernel_ordering.nr

```diff
@@ -43,25 +50,56 @@ impl PrivateKernelInputsOrdering {
 
         // match reads to commitments from the previous call(s)
         for rr_idx in 0..MAX_READ_REQUESTS_PER_TX {
-            let read_request = read_requests.get_unchecked(rr_idx).value;
-            let read_commitment_hint = self.read_commitment_hints[rr_idx];
-            //TODO(David): Shouldn't this just be a uint64?
-            let hint_pos = read_commitment_hint as u64;
-
-            if (read_request != 0) {
-                let mut match_pos = MAX_NEW_COMMITMENTS_PER_TX as u64;
-                if (hint_pos < MAX_NEW_COMMITMENTS_PER_TX as u64) {
-                    match_pos = if read_request == new_commitments.get_unchecked(hint_pos as Field).value { hint_pos } else { match_pos };
-                }
-            
-                assert(match_pos != MAX_NEW_COMMITMENTS_PER_TX as u64, "read request is transient but does not match any commitment");
+            let read_request = read_requests.get_unchecked(rr_idx);
+            let read_commitment_hint = self.read_commitment_hints[rr_idx] as u64;
+
+            if (read_request.value != 0) {
+                let commitment = new_commitments.get_unchecked(read_commitment_hint as Field);
+                assert_eq(read_request.value, commitment.value, "Hinted commitment does not match read request");
+                assert(read_request.counter > commitment.counter, "Read request counter must be greater than commitment counter");
             }
         }
 
         // Empty out read requests after matching them to commitments
         public_inputs.end.read_requests = BoundedVec::new(SideEffect::empty());
     }
 
+    fn assert_sorted_counters<T, N>(original: [T; N], sorted: [T; N], indexes: [u32; N]) where T: Eq + Ordered + Empty {
+        let mut seen = [false; N];
+        let mut prev_counter = Option::none();
+        let mut prev_was_empty = false;
+
+        for i in 0..N {
+            let item = sorted[i];
+
+            let index = indexes[i];
+            assert(index < N, "Index out of bounds");
+            assert(!seen[index], "Duplicate index");
+            seen[index] = true;
```

Comment:
Don't have to use `seen` here. Since the counter of every item must be larger than the previous counter, it's not possible to refer to the same index more than once.

Reply:
True! Since I made it like "first make sure it's a permutation, then check the permutation rules" I didn't realize that part of the permutation checks were superceded :D

Reply:
Oh I think we have to keep the seen check to ensure that no deletion happens, since counters don't have to be increasing for the empty ones
```rust
    fn assert_sorted_counters_no_seen<T, N>(
        original: [T; N],
        sorted: [T; N],
        indexes: [u32; N]
    ) where T: Eq + Ordered + Empty {
        let mut prev_counter = Option::none();
        let mut prev_was_empty = false;

        for i in 0..N {
            let item = sorted[i];

            let index = indexes[i];
            assert(index < N, "Index out of bounds");

            assert(item.eq(original[index]), "Not equal");

            let counter = item.counter();
            let is_empty = is_empty(item);

            if prev_was_empty {
                assert(is_empty, "Empty items must be at the end");
            } else if !is_empty & prev_counter.is_some() {
                assert(counter > prev_counter.unwrap(), "Not sorted");
            }

            prev_was_empty = is_empty;
            prev_counter = Option::some(counter);
        }
    }
    
    // This passes ):
    #[test(should_fail)]
    unconstrained fn check_cannot_delete() {
        let original = [
            SideEffect { value: 1, counter: 1 }, SideEffect { value: 2, counter: 2 }, SideEffect::empty()
        ];
        let sorted = [SideEffect { value: 1, counter: 1 }, SideEffect::empty(), SideEffect::empty()];
        let indexes = [0, 2, 2];

        assert_sorted_counters_no_seen(original, sorted, indexes)
    }
}
```

Reply:
We could count non empty items on original and sorted, but at that point we might not be saving on constraints

Reply:
Updated after some discussions on the best approach to optimize it!